lib/msal-node/src/client/ClientCredentialClient.ts

```diff
@@ -58,7 +58,20 @@ export class ClientCredentialClient extends BaseClient {
 
         const cachedAuthenticationResult =
             await this.getCachedAuthenticationResult(request);
+
         if (cachedAuthenticationResult) {
+            // if the token is not expired but must be refreshed; get a new one in the background
+            if (this.serverTelemetryManager?.getCacheOutcome() === CacheOutcome.REFRESH_CACHED_ACCESS_TOKEN) {
+                this.logger.info(
+                    "ClientCredentialClient:getCachedAuthenticationResult - Cached access token's refreshOn property has been exceeded'. It's not expired, but must be refreshed."
+                );
+
+                // return the cached token, don't wait for the result of this request
+                const accessTokenRefresh = true;
+                this.executeTokenRequest(request, this.authority, accessTokenRefresh);
```

Comment:
@Robbie-Microsoft @derisen - what happens with background request here? Does NodeJS keep the thread alive until the request finishes and the promise returns?

Reply:
yep

Reply:
Nodejs has a single thread (the event loop) to execute JS code (+k worker threads, but used mainly for things like GC and disk access unless explicitly created). So it will invoke the `executeTokenRequest` but will not wait for it to complete and continue doing other things

Reply:
@derisen - I get that it will just go to line 73, but what happens to the `executeTokenRequest` ? It's a long running async operation, will the current request wait for it? 

Should we use https://nodejs.org/api/worker_threads.html#worker-threads to schedule this work? If not, why not?

Reply:
> Point 5 might or might not happen before the token acquisition (at step 3) is complete. Does NodeJS guarantee to not kill the thread and wait for step 3 to complete? (ASP.NET Core does this btw).

That's the thing though, there's only one thread (event loop) and it's alive as long as the app runs. In other words, no new thread is created for the network call here -it will be queued in the event loop.

> From looking at SO, I see that most ppl point to use https://nodejs.org/api/worker_threads.html#worker-threads to schedule step 3.

Worker threads are recommended for CPU intensive tasks. As mentioned in the docs, they don't provide much advantage for IO operations.

So I believe it's guaranteed, but @tnorling might have more insight on this.

Reply:
The node thread will not close until the event loop is empty (unless the process is explicitly killed)

Reply:
Probably good to have test coverage for this scenario though

Reply:
So, we are updating the `accessToken` in the background and is the below use case valid in this case: 

1. access_token fetched
2. App uses the access_token 
3. background refresh initiated
4. App calls atSilent, gets the old token, which can be expired between 3 and 4 - fails get interaction needed error
5. background fetch successful
6. App retries after step 4 and now gets a valid token??

Reply:
@sameerag - I think MSAL will **either** kick off a background fetch (refresh logic) **or** treat the cached token expired. Both cannot happen at the same time, since it's the call to `atSilent` / `atClientCreds` that initiates the entire logic.

Reply:
Thanks for weighing in on this @tnorling @sameerag! 

@Robbie-Microsoft - please see https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick and consider doing a manual test first to confirm what @tnorling said. You can automate a test by adding a 1-2s timeout in the the refresh logic